Add discovery and cluster formation topic (#11416)
opensearch-project/documentation-website: _tuning-your-cluster/discovery-cluster-formation/bootstrapping.md

---
layout: default
title: Cluster bootstrapping
parent: Discovery and cluster formation
nav_order: 40
---

# Cluster bootstrapping

> When starting an OpenSearch cluster for the very first time, you must explicitly define the initial set of cluster-manager-eligible nodes that will participate in the first cluster manager election. This process is called _cluster bootstrapping_ and is critical for preventing split-brain scenarios during initial cluster formation.

Reviewer: I think we can omit the split brain part here as it is just the bootstrapping that prevents it but the entire consensus algorithm makes sure that it does not happen.

Cluster bootstrapping is required in the following situations:

- Starting a brand-new cluster for the very first time

> - No existing cluster state exists on any node

Reviewer 2: ```suggestion
- No existing cluster state exists on any node.
```

> - Initial cluster manager election needs to take place

Reviewer 2: ```suggestion
- Initial cluster manager election needs to take place.
```

Bootstrapping is not required in the following situations:

> - Nodes joining an existing cluster - they get configuration from the current cluster manager

Reviewer 2: ```suggestion
- Nodes joining an existing cluster: They get their configuration from the current cluster manager.
```

> - Cluster restarts - nodes that have previously joined a cluster store the necessary information

Reviewer 2: ```suggestion
- Cluster restarts: Nodes that have previously joined a cluster store the necessary information.
```

**Correct configuration**:

```yaml
# Node 1
cluster.initial_cluster_manager_nodes:
  - cluster-manager-1
  - cluster-manager-2
  - cluster-manager-3

# Node 2
cluster.initial_cluster_manager_nodes:
  - cluster-manager-1
  - cluster-manager-2
  - cluster-manager-3

# Node 3
cluster.initial_cluster_manager_nodes:
  - cluster-manager-1
  - cluster-manager-2
  - cluster-manager-3
```
{% include copy.html %}

**Incorrect configuration**:

```yaml
# Node 1 – different list
cluster.initial_cluster_manager_nodes:
  - cluster-manager-1
  - cluster-manager-2

# Node 2 – different list
cluster.initial_cluster_manager_nodes:
  - cluster-manager-2
  - cluster-manager-3
```

When nodes have inconsistent bootstrap lists, multiple independent clusters may form.

### Exact name matching

Node names in the bootstrap configuration must exactly match each node's `node.name` value.

**Common naming issues**:

* If a node's name is `server1.example.com`, the bootstrap list must also use `server1.example.com`, not `server1`.
* Node names are case-sensitive.

> * The names must match exactly, with no added characters or whitespace.

Reviewer 2: ```suggestion
* The names must match exactly, with no added characters or white space.
```

- `discovery.seed_providers`
- `discovery.seed_hosts`
- `cluster.initial_cluster_manager_nodes`

### Auto-bootstrapping limitations

Auto-bootstrapping is intended only for development. Do not use it in production because:

- Nodes may not discover each other quickly enough, leading to delays.

- Network conditions can cause discovery to fail.

- Behavior can be unpredictable and is not guaranteed.

- There is a risk of forming multiple clusters, resulting in split-brain scenarios.

## Troubleshooting bootstrap issues

If you accidentally start nodes on different hosts without proper configuration, they may form separate clusters. You can detect this by checking cluster UUIDs:

```bash
curl -X GET "localhost:9200/"
```
{% include copy.html %}

If each node reports a different `cluster_uuid`, they belong to separate clusters. To correct this and form a single cluster, use the following steps:

1. Stop all nodes.
2. Delete all data from each node's data directory.
3. Configure proper bootstrap settings.
4. Restart all nodes and verify single cluster formation.

## Bootstrap verification

> After starting your cluster, verify successful bootstrap using the [monitoring commands]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/#monitoring-discovery-and-cluster-formation) for checking cluster health and formation:

Reviewer 2: ```suggestion
After starting your cluster, verify successful bootstrapping using the [monitoring commands]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/#monitoring-discovery-and-cluster-formation) for checking cluster health and formation:
```

- Verify cluster health status and node count.
- Confirm that one node is elected as cluster manager.
- Ensure that all nodes report the same cluster UUID.

## Related documentation

> - [Voting configuration management]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/voting-configuration/): How OpenSearch manages voting after bootstrap

Reviewer 2: Same here. "after bootstrapping"?

> - [Voting configuration management]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/voting-configuration/): How OpenSearch manages voting after bootstrap

Reviewer 2: ```suggestion
- [Voting configuration management]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/voting-configuration/): How OpenSearch manages voting after bootstrapping
```
